python/lsst/meas/base/recalibrateExposure.py

```diff
@@ -0,0 +1,321 @@
+            for name in calibs:
+                if calibs[name] is None:
+                    continue
+                calibs[name] = calibs[name].pop()
+        return calibs
+
+    def readCalibs(self, dataRef):
+        """Read all inputs
+
+        This includes the exposure, photometric and astrometric calibrations,
+        and sky correction.
+
+        This method is intended for use with the Gen2 middleware only.
+
+        Parameters
+        ----------
+        dataRef : `lsst.daf.persistence.ButlerDataRef`
+            Butler data reference.
+
+        Returns
+        -------
+        results : `lsst.pipe.base.Struct`
+            Inputs that were read, including the following elements:
+
+            ``photoCalib``
+                Photometric calibration (`lsst.afw.image.PhotoCalib`), or
+                ``None`` if not ``doApplyExternalPhotoCalib``.
+            ``skyWcs``
+                Astrometric calibration (`lsst.afw.geom.SkyWcs`), or ``None``
+                if not ``doApplyExternalSkyWcs``.
+            ``skyCorr``
+                Sky correction to apply, or ``None`` if not ``doApplySkyCorr``.
+        """
+        photoCalib = None
+        if self.config.doApplyExternalPhotoCalib:
+            photoCalib = dataRef.get(f"{self.config.externalPhotoCalibName}_photoCalib")
+
+        skyWcs = None
+        if self.config.doApplyExternalSkyWcs:
+            skyWcs = dataRef.get(f"{self.config.externalSkyWcsName}_wcs")
+
+        skyCorr = None
+        if self.config.doApplySkyCorr:
+            skyCorr = dataRef.get("skyCorr")
+
+        return Struct(photoCalib=photoCalib, skyWcs=skyWcs, skyCorr=skyCorr)
+
+    def run(self, exposure, photoCalib=None, skyWcs=None, skyCorr=None, rescale=True):
+        """Apply calibrations to an exposure
+
+        The result is an exposure with units of nJy, unless ``rescale``, in
+        which case the mean calibration is used.
+
+        Parameters
+        ----------
+        exposure : `lsst.afw.image.Exposure`
+            Exposure to recalibrate. Will be modified.
+        photoCalib : `lsst.afw.image.PhotoCalib`, optional
+            Photometric calibration, or ``None`` if no photometric calibration
+            is to be applied.
+        skyWcs : `lsst.afw.geom.SkyWcs`, optional
+            Astrometric calibration, or ``None`` if no astrometric calibration
+            is to be applied.
+        skyCorr : `lsst.afw.math.BackgroundList`, optional
+            Sky correction, or ``None`` if no sky correction is to be applied.
+        rescale : `bool`
+            Rescale to mean zero-point?
+
+        Returns
+        -------
+        exposure : `lsst.afw.image.Exposure`
+            Recalibrated exposure.
+        """
+        if photoCalib is None:
+            photoCalib = exposure.getPhotoCalib()
+        exposure.maskedImage = photoCalib.calibrateImage(exposure.maskedImage,
+                                                         self.config.includePhotoCalibVar)
```

Comment:
I think this step should be optional - it's expensive, and we don't need it if we're only going to do catalog operations (e.g. `ForcedPhotCcd`!), for which attaching a spatially-varying `PhotoCalib` to the `Exposure` is sufficient.